keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1181 @@
+
+### Cross-Namespace ApplySet Contents
+
+When querying for ApplySet contents, an ApplySet could contain cluster-scoped resources or could contain resources in other namespaces.  Querying for this content is generally going to require more permissions and be slower, so we would like to avoid over-querying here.
+
+Best practice is likely to avoid ApplySets spanning namespaces.  However, sometimes this is unavoidable - particularly when managing cluster-scoped objects - and the “plumbing” tooling cannot enforce this restriction.
+
+Where a GK is known to be part of the ApplySet and is cluster-scoped, we should naturally query for those objects at cluster scope; any permission problems here should be surfaced as errors.  When a tool cannot determine the list of GKs for an ApplySet, it may support “discovery”, likely warning that the ApplySet does not define a list of GKs, and then attempt to perform cluster-scoped queries, likely warning if there are insufficient permissions.
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> Should kubectl ever fall back to this? Will it have a separate subcommand for it?
+>
+> <<[/UNRESOLVED]>>
+
+For GKs that are namespace-scoped, we would normally expect those to be part of an ApplySet object in the same namespace.  We define an additional annotation however for cross-namespace ApplySets:  `applyset.k8s.io/additional-namespaces`.  The value of this annotation is a comma-separated list of the names of the namespaces (other than the ApplySet namespace) in which objects are found, for example `default,kube-system,ns1,ns2`.  Note that there is no need to include this specifically for cluster-scoped objects, as those are covered by the group-kind list.  We reserve the empty value.  If this annotation is present, the tooling will query namespace-scoped resources in those namespaces in addition to the namespace of the ApplySet object (if any).  If this annotation is not present on a namespace-scoped ApplySet parent object, the tooling will query namespace-scoped resources only in the same namespace as the ApplySet parent object.  If the annotation is not present on a cluster-scoped ApplySet parent object, the tooling will not query namespace-scoped resources at all and should output an error if given namespace-scoped GKs.
+
+As with `applyset.k8s.io/contains-group-kinds`, this list of namespaces must be sorted alphabetically, and should be minimal (ideally other than during apply and prune operations).
+
+As cross-namespace ApplySets are not particularly encouraged, we do not currently optimize this further.  In particular, we do not specify the GKs per namespace.  We can add more annotations in future should the need arise.
+
```

Comment:
I'm missing a description of a case with a mixture of namespace-scope and cluster-scoped resources. How that case is going to be handled?

Reply:
Justin added a paragraph about this below; PTAL and un-resolve this convo if needed.